In [325]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.feature_extraction.text import TfidfVectorizer

In [326]:
df = pd.read_csv('mail_data.csv')

No Null Values

In [327]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Category  5572 non-null   object
 1   Message   5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB


Remove Duplicates

In [328]:
# df.drop_duplicates(keep='first',inplace=True)
print(f"the total number of duplicated found in this dataset are {df[df.duplicated(subset = 'Message',keep='first')].shape[0]}")

the total number of duplicated found in this dataset are 415


In [329]:
df.drop_duplicates(keep='first',inplace=True)

In [330]:
df.reset_index(drop=True,inplace=True)

In [331]:
from sklearn.preprocessing import OrdinalEncoder
oe = OrdinalEncoder(categories=[['spam','ham']])
df['Category'] = oe.fit_transform(df[['Category']])

In [332]:
df['Category'] = df['Category'].astype(int)

In [333]:
x = df.iloc[:,[1]]
y = df.iloc[:,[0]]

In [334]:
x

Message
0     Go until jurong point, crazy.. Available only ...
1                         Ok lar... Joking wif u oni...
2     Free entry in 2 a wkly comp to win FA Cup fina...
3     U dun say so early hor... U c already then say...
4     Nah I don't think he goes to usf, he lives aro...
...                                                 ...
5152  This is the 2nd time we have tried 2 contact u...
5153               Will ü b going to esplanade fr home?
5154  Pity, * was in mood for that. So...any other s...
5155  The guy did some bitching but I acted like i'd...
5156                         Rofl. Its true to its name

[5157 rows x 1 columns]

Convert the X which is raw data into vector using FeatureExtraction

In [335]:
tv = TfidfVectorizer(min_df=1,stop_words='english',lowercase=True)
x = tv.fit_transform(df['Message'])
# Stop words are a an the which are sort of useless in these case , min_df = whenever the given value is less than 1 then its ignored as it cannot play imp role in prediction 
# max feature select = 5000 only select 5000 vdimension 

(5157, 8440) is the shape of vector where each and every row message is given as an combo of 8440 words

In [336]:
x

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 39457 stored elements and shape (5157, 8440)>

get_features_name_out gives us all 8440 feature name

In [337]:
tv.get_feature_names_out()

array(['00', '000', '000pes', ..., 'èn', 'ú1', '〨ud'], dtype=object)

In [338]:
from sklearn.model_selection import train_test_split
x_train , x_test , y_train , y_test = train_test_split(x,y,random_state=42,test_size=0.2)

In [339]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(x_train,y_train)

/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/utils/validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression()

In [340]:
lr.score(x_test,y_test) , lr.score(x_train,y_train) 
# No Overfitting

(0.937984496124031, 0.9621818181818181)

In [341]:
print(f"The accuracy of Logistic Regression Model is {round(100*lr.score(x_test,y_test),3)}%")

The accuracy of Logistic Regression Model is 93.798%


Lets  Take an example and see what result we are going to get , we are going to convert the sentence into sparced_matrix usimg transform  , no need to use fit as it already have been trained on it

In [342]:
input_mail  = ['spam,"XXXMobileMovieClub: To use your credit, click the WAP link in the next txt message or click here>> http://wap. xxxmobilemovieclub.com?n=QJKGIGHJJGCBL"']
input_mail_vectorisation = tv.transform(input_mail)

In [343]:
lr.predict(input_mail_vectorisation)

array([0])